### PR TITLE
Stream API for AbstractNavigation

### PR DESCRIPTION
**Pull request by julianseeger**
depends on https://github.com/Qabel/qabel-core/pull/524

### CODE REVIEW

**Comment by julianseeger** [comment]
this is imho done with https://github.com/Qabel/qabel-core/pull/526/commits/b13b42b8144d3334c2b669dccf8250ea261caddc

File: box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt [hunk 1]
```diff
@@ -43,13 +43,18 @@ abstract class AbstractNavigation(
     private var autocommit = true
     private var autocommitDelay = DEFAULT_AUTOCOMMIT_DELAY
     private var lastAutocommitStart: Long = 0
+    var time: () -> Long = { System.currentTimeMillis() }
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: Why is is not a method?

File: box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt [hunk 2]
```diff
@@ -159,35 +157,43 @@ abstract class AbstractNavigation(
...
+    override fun upload(name: String, file: InputStream, size: Long, listener: ProgressListener?)
+        = uploadStream(file, name, size, time(), listener)
 
     @Synchronized @Throws(QblStorageException::class)
     override fun overwrite(name: String, file: File, listener: ProgressListener?): BoxFile
         =  uploadFile(name, file, null, listener)
 
     @Throws(QblStorageException::class)
     private fun uploadFile(name: String, file: File, expectedFile: BoxFileState?, listener: ProgressListener?): BoxFile {
+        val mtime: Long
```
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: try/catch is an expression:
`val a: Int? = try { parseInt(input) } catch (e: NumberFormatException) { null }`

File: box/src/test/java/de/qabel/box/storage/BoxVolumeTest.java [hunk 3]
```diff
@@ -140,6 +144,28 @@ public void testDeleteFile() throws QblStorageException, IOException {
         nav.download(boxFile);
     }
 
+    @Test
+    public void uploadsStreams() throws Exception {
+        InputStream in = new ByteArrayInputStream("testContent".getBytes());
+        Long size = 11L;
+
+        DefaultIndexNavigation nav = volume.navigate();
+        nav.setTime(new Function0<Long>() { @Override public Long invoke() { return 1234567890L; }}); // imagine lambda
```
Comment (audax) [inline box/src/test/java/de/qabel/box/storage/BoxVolumeTest.java]: this is why it is not a method, I see…

It looks weird, though.

File: box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt (hunk 1 above)
Comment (julianseeger) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: ? it's a callback ... Imagine this as a 
```JAVA
@FunctionalInterface
class DefaultTimeProvider implements TimeProvider {
    @Override
    public Long getTime() {
        return System.currentTimeMillis();
    }
}
```

File: box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt (hunk 2 above)
Comment (julianseeger) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: hats actually more code ;) but I'll try

**Comment by julianseeger** [comment]
AFTER this PR, I'm going to convert the test class to kotlin ... it looks this ugly because of the missing lambdas / functional interfaces at this level. But konverting it inside this PR would obfuscate too much

File: box/src/test/java/de/qabel/box/storage/BoxVolumeTest.java (hunk 3 above)
Comment (julianseeger) [inline box/src/test/java/de/qabel/box/storage/BoxVolumeTest.java]: image
btw

Comment (audax) [inline box/src/test/java/de/qabel/box/storage/BoxVolumeTest.java]: That looks better :D

File: box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt (hunk 1 above)
Comment (audax) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: A rename to "timeProvider" would be enough to make the intent clear, imo.

Comment (julianseeger) [inline box/src/main/java/de/qabel/box/storage/AbstractNavigation.kt]: true. `time` looked good in the test so I thought that was a good idea back then o.O

**Pull request merged**